Child and descendant selectors

cjohansen opened the pull request:
This pull request is based on the work in #2. This one allows you to distinguish between descendant selectors and direct children:

```clojure
[:div :a] ;; Descendant selector, e.g. "any a under a div"
[:div :> :a] ;; Children selector, e.g. "any a directly child to a div"
```

This is technically a breaking change, as the current implementation could be considered buggy - it considers `[:div :a]` to be a direct child selector. This breaks with expectations if you're familiar with CSS, but changing it could change the behavior of existing code... Not sure what the best way to approach this is, but at the very least it would be good to have support for descendant selectors.

My gut feeling is that this change - while breaking - is ok, since my guess is that people are already using selectors expecting this "new" behavior. But I'm not sure.

cjohansen committed:
Enable both descendant/child selectors, match CSS selector semantics

magnars commented:
With regards to the breaking change, maybe we could consider moving to a different namespace?

One quick idea, if we're considering switching namespace: If we change to quoted lookup paths, we could more closely resemble the CSS it is inspired from. Things like this would be possible:

```
'[div.foo a > :first-child]
```

Any thoughts?

cjohansen commented:
Since making the pull request I've had the same idea for using quoting 😄 I like your suggestion a lot!

cjohansen committed:
Solve the child/descendant conundrum with a new namespace

cjohansen commented:
I've added the `html5-walker.walker` namespace and updated the README accordingly. The `html5-walker.core` namespace works as before, and the new namespace has the more CSS-aligned behavior.

Re: symbols. As I've discovered at least once before (Sanity queries) - this already works out of the box, since symbols are named. So really symbols, keywords and strings may be used interchangeably. You will have to use strings for attribute selectors, but for everything else I think symbols look the best, so I updated the README to use those.

magnars commented:
That's pretty neat, I must say. Cheers!

Would you mind rebasing onto master? There seems to be conflicts after PR #2 was merged.

cjohansen committed:
Lean on the new walker namespace from legacy core

cjohansen commented:
Done!

magnars commented:
Now available as `html5-walker {:mvn/version "2023.10.22"}` 👍
